text/0695-default-args.md

```diff
@@ -0,0 +1,556 @@
+components.
+
+Allowing argument values to provide defaults, _without_ allowing them to
+override local values, is more like the way default values in JavaScript
+functions actually work:
+
+```js
+function foo(a = 1, b = 2) {
+  let localState = 3;
+
+  return a + b + localState;
+}
+
+foo(4, 5); // 12
+```
+
+Here we can clearly see which values are internal and external, and also which
+values are potentially defaulted and what they're defaulted to. Default values
+in components would work the same way, without the ability to mix with local
+component state, and with default values clearly visible for users to see.
+
+There are a number of different possible high level APIs for defaults that we
+could introduce to components. For instance, default args could be specified as
+a static class field on Glimmer components:
+
+```js
+import Component from '@glimmer/component';
+
+export default class MyComponent extends Component {
+  static defaultArgs = {
+    a: 1,
+    b: 2,
+  };
+}
+```
+
+But this would not work well for Template-only components. All potential designs
+will also likely be influenced by the upcoming changes to introduce Template
+Imports, which _itself_ has not been entirely figured out. For instance,
+default could be associated by passing them into the template definition:
+
+```js
+const defaultArgs = {
+  a: 1,
+  b: 2,
+};
+
+// Using the hbs`` style
+export default hbs({ defaultArgs })`
+  Hello, world!
+`;
+
+// Using a possible custom syntax
+<template @defaultArgs={{defaultArgs}}>
+  Hello, world!
+</template>
+```
+
+As such, this RFC is not proposing a high level API to be added directly.
+Instead, we propose adding a `getDefaultArgs` hook to component managers in
+order to provide a standard place for argument defaults to be defined. This hook
+is meant to be a low level primitive, which can be used by advanced users and
+addon authors to explore high level APIs and ways to add default arguments.
+
+## Detailed design
+
+The `getDefaultArgs` hook will be added to each of the component manager
+interface with the following signature:
+
+```ts
+interface DefaultArgumentValue =
+  | string
+  | number
+  | boolean
+  | null
+  | undefined
+  | readonly Array<unknown>
+  | ReadOnly<unknown>
+  | () => unknown;
+
+interface DefaultArguments {
+  named?: Record<string, DefaultArgumentValue>;
+  positional?: DefaultArgumentValue[];
+}
+
+interface ComponentManager {
+  getDefaultArgs(definition: object): DefaultArguments | null;
+}
+```
+
+This hook receives the definition of the component as its first and only
+argument, and returns an object containing default argument values. Valid
+default argument values are either:
+
+1. Primitive JavaScript values
+2. Frozen arrays or object
+3. Functions which return a default value
+
+The default value will be used for the corresponding named and positional
+arguments if the argument that was passed in is either `null` or `undefined`.
```

Comment:
What are the benefits to treat `null` the same as `undefined`? It seems to limit the component APIs significantly.

Reply:
The main idea was just keeping it consistent between the way `null` and `undefined` are handled when interpolating in templates, so the rules are consistent in templates. I agree this could be limiting though, and in JS for instance, `??` is different than argument defaulting. `??` will coalesce if the value is `null` or `undefined`, whereas argument defaulting will use the default value _only_ if it is `undefined`.

Are there motivating use cases you can think of for when users would want to be able to pass `null`, and not trigger a default? It would be helpful if we could show such use cases in the motivation

Reply:
Not a specific use case, but I believe these are just semantically different. I may want to tell a component, the value must be initialized with nothing.

Reply:
I find it difficult to come up with a realistic use case. I can image many components, which support `null` as a value for one of their arguments. So please excuse if that example may sound a little bit made up.

Let's imagine a `<Switch>` component, which three states: `true`, `false` or `null`. The component may be used in a poll to select the answers a) yes, b) no, c) I don't know. 

The component has a `@value` argument, which allows a consumer to pass in the state. For unknown reasons the developer of the component decides that the argument is optional and that `true` is a good default for it.

If the argument default primitives do not distinguish between `null` and `undefined`, this use case would not be supported. The default value would be used even if the consumer explicitly invokes the component with `@value={{null}}`.

Reply:
We use null to opt out of using a default component name. This could probably be better modeled with named blocks though.